controlplane/kubeadm/controllers/helpers.go

```diff
@@ -256,16 +255,36 @@ func (r *KubeadmControlPlaneReconciler) generateMachine(ctx context.Context, kcp
 	return nil
 }
 
-// machinesNeedingRollout return a list of machines that need to be rolled out.
-func (r *KubeadmControlPlaneReconciler) machinesNeedingRollout(ctx context.Context, c *internal.ControlPlane) internal.FilterableMachineCollection {
-	// Ignore machines to be deleted.
-	machines := c.Machines.Filter(machinefilters.Not(machinefilters.HasDeletionTimestamp))
-
-	// Return machines if they are scheduled for rollout or if with an outdated configuration.
-	return machines.AnyFilter(
-		// Machines that are scheduled for rollout (KCP.Spec.UpgradeAfter set, the UpgradeAfter deadline is expired, and the machine was created before the deadline).
-		machinefilters.ShouldRolloutAfter(c.KCP.Spec.UpgradeAfter),
-		// Machines that do not match with KCP config.
-		machinefilters.Not(machinefilters.MatchesKCPConfiguration(ctx, r.Client, c.KCP)),
-	)
+func (r *KubeadmControlPlaneReconciler) getInfraObjects(ctx context.Context, machines internal.FilterableMachineCollection) (map[string]*unstructured.Unstructured, error) {
+	result := map[string]*unstructured.Unstructured{}
+	for _, m := range machines {
+		infraObj, err := external.Get(ctx, r.Client, &m.Spec.InfrastructureRef, m.Namespace)
+		if err != nil {
+			if apierrors.IsNotFound(errors.Cause(err)) {
+				continue
+			}
+			return nil, errors.Wrapf(err, "failed to retrieve infra obj for machine %q", m.Name)
+		}
+		result[m.Name] = infraObj
+	}
+	return result, nil
+}
+
+func (r *KubeadmControlPlaneReconciler) getMachineConfigs(ctx context.Context, machines internal.FilterableMachineCollection) (map[string]*bootstrapv1.KubeadmConfig, error) {
```

Comment:
nit: I think we should be more explicit with the naming here, I was a bit confused by what a MachineConfig was initially.
```suggestion
func (r *KubeadmControlPlaneReconciler) getKubeadmConfigs(ctx context.Context, machines internal.FilterableMachineCollection) (map[string]*bootstrapv1.KubeadmConfig, error) {
```

Reply:
Yeah I was pulling this from a variable we had called `machineConfig` and wasn't sure if that was common terminology for these